lib/grp.gi

```diff
@@ -465,11 +465,44 @@ local p, hom, reps, as, a, b, ap, bp, ab, ap_bp, ab_p, g, h, H, N;
   elif Size(G) < p^p * Size(Agemo(G,p)) then
     # see [Hal36, Theorem 2.3], [Hup67, Satz III.10.13]
     return true;
-  elif Index(DerivedSubgroup(G),Agemo(DerivedSubgroup(G),p)) < p^(p-1) then
+  elif Index(DerivedSubgroup(G), Agemo(DerivedSubgroup(G),p)) < p^(p-1) then
     # see [Hal36, Theorem 2.3], [Hup67, Satz III.10.13]
     return true;
   fi;
 
+  # We now use Proposition 2 from A. Mann, "Regular p-groups. II", 1972, DOI
+  # 10.1007/BF02764891, which states: If N is a central elementary abelian
+  # subgroup of order p^2, such that G/M is regular for all M with 1<M<N, then
+  # G is regular. The reverse implication also holds as all sections of a
+  # regular p-group are again regular.
+  #
+  # Such a subgroup exists if and only if the center of G is not cyclic.
+  #
+  # As a heuristic, we only apply this criterion if the index of the center in
+  # G is not too small, as otherwise a brute force search is faster.
+  #
+  # Note: the book Y. Berkovich, "Groups of Prime Power Order, Volume 1", 2008
+  # states a stronger version of this as Corollary 7.7, where it is basically
+  # claimed that it suffices to check just two subgroups M of N. This result
+  # is attributed to the above paper by Mann, but I can't find it in there,
+  # and it also simply is wrong: for example, the direct product of
+  # SmallGroup(3^5,22) and SmallGroup(3^5,22) has a center if order p^2 = 9,
+  # which contains four subgroups M of order p = 3. For two of those the
+  # corresponding quotient G/M is regular, and for the other two it is not.
```

Comment:
@jackschmidt @ThomasBreuer FYI I've now added this paragraph

Reply:
Looks good. Typo in line 489, if -> of